### PR TITLE
 Rework on Augmented Lagrangian Multiplier for contact problems

### LINKED ISSUE #9677: Augmented Lagrangian Multiplier Method for Contact Problem
@bwspenc 

### Description of the enhancement or error report
This work implemented an augmented lagrangian multiplier (ALM) method to enforce the contact constraint. Compared to the penalty method, there are a couple of strengths for ALM method. ALM method is variational consistent and therefore can hold the optimal convergence rate. Moreover, ALM method is 

### Identified impact
The following classes are changed
Transient (Only add two lines)
Constraint (Add three virtual functions)
MechanicalContactConstraint (Add three functions, modify computecontactforce, residual, jacobian)
FEProblemBase(Add three functions)
NonlinearSystemBase (Add three functions)

Add one class 
ContactAugLagMulProblem

Add three tests
modules/combined/tests/contact_verification/hertz_cyl/half_symm_q4/hertz_cyl_half_1deg_template2_sm_alm.i
modules/combined/tests/sliding_block/sliding/constraint/frictional_02_alm.i
modules/combined/tests/sliding_block/sliding/constraint/frictionless_alm.i

### PR DESCRIPTION
**Pull request by jiangwen84**
Yingjie (@yingjie1989) has finished his internship at INL and went back to school. He submitted the PR (#9678) which still has some issues with rebasing and failure tests. I took his branch and opened this new PR to resolve issues and address comments. 

His original PR information is below

- After

 nonlinear solve converged, check the convergence criterion. If satisfied, accept the solution and go to the next step. If not, update the lagrangian multiplier and repeat the nonlinear solve

- test

modules/combined/test/tests/contact_verification/hertz_cyl/half_symm_q4/hertz_cyl_half_1deg_template2_sm_alm.i
modules/combined/test/tests/sliding_block/sliding/constraint/frictional_02_alm.i
modules/combined/test/tests/sliding_block/sliding/constraint/frictionless_alm.i

Close #9677

### CODE REVIEW

**Comment by permcody** [comment]
Oh, I like the new size warning! 

It looks like you are adding new regression tests that contain a thousand steps. Is this necessary or can you reduce the size? Even if you need a 1000 steps to fully test the physics, you could likely output on a subset of the steps (1 / 10 or 100) for regression testing and reduce the sizes substantially. It all really adds up over time. Take a second look before we commit and see if you can shrink this a little.

**Comment by jiangwen84** [comment]
It does not need 1000 steps at all :-). It was a typo. I will fix that and try to reduce other tests.

**Comment by jiangwen84** [comment]
I will squash all the commits into one when the review is done.

**Comment by permcody** [comment]
@jiangwen84 - Looks like this PR just needs documentation added.

**Comment by jiangwen84** [comment]
I added an empty document for ALM problem. I will complete it later.

**Comment by permcody** [comment]
Please define EXACTLY when you mean by later...

**Comment by jiangwen84** [comment]
@permcody I am not sure what the comments I will have. It is likely that some comments might ask me to change the way I code for the "ALM problem". So I want to wait until I get some comments and then finalize the documentation.

**Comment by jiangwen84** [comment]
@permcody,  the module tests failed and the info is below. 

1885 passed, 202 skipped, 0 pending, 0 failed, FATAL PARSER ERROR
ERROR: exiting with code 1

I do not know what causes this, do you have any suggestions?

**Comment by jiangwen84** [comment]
@permcody I think it was a bug in the parser check. In the test file, I had 

skip = 'waiting for alm merge's comments'. 

This gave me error saying 'expected = ... after s but get c...'

I think it does not like the 's after the merge. So I switch to 

skip = 'waiting for alm merge'

And it works now.

**Comment by permcody** [comment]
Wen, It's working perfectly. Your string contains a single quote character but you used that as your string delimiter. In Python you can use either single quote or double quote characters (the new HIT parser works the same way) to delimit your string. You can't reuse the same character within the string though without escaping it. The easy fix here would be to use double quotes for your message.

**Comment by jiangwen84** [comment]
I got it. It sounds great:-)

**Comment by jiangwen84** [comment]
I just squash all the commits.

**Comment by bwspenc** [comment]
@jiangwen84 I just merged Al's code, which causes a couple of conflicts with your code. Can you please rebase and fix the conflicts?

File: modules/combined/test/tests/contact/pressureAugLagNew.i
Comment (bwspenc) [inline modules/combined/test/tests/contact/pressureAugLagNew.i]: It doesn't look this file is used in a 'tests' file.

File: modules/combined/test/tests/contact/pressureAugLagNew.i
Comment (bwspenc) [inline modules/combined/test/tests/contact/pressureAugLagNew.i]: If you're using 'lu', you'll need to limit the test to run in serial only.

File: modules/combined/test/tests/contact/pressureAugLagNew.i
Comment (bwspenc) [inline modules/combined/test/tests/contact/pressureAugLagNew.i]: remove commented out line

File: modules/contact/include/AugmentedLagrangianContactProblem.h
Comment (bwspenc) [inline modules/contact/include/AugmentedLagrangianContactProblem.h]: Class to manage nested solution for augmented Lagrange contact.

File: modules/contact/include/AugmentedLagrangianContactProblem.h
Comment (bwspenc) [inline modules/contact/include/AugmentedLagrangianContactProblem.h]: override

File: modules/contact/include/AugmentedLagrangianContactProblem.h
Comment (bwspenc) [inline modules/contact/include/AugmentedLagrangianContactProblem.h]: Add Doxygen comments to this header file.

File: modules/contact/src/AugmentedLagrangianContactProblem.C
Comment (bwspenc) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: You also need to handle the case where nc isn't a MechanicalContactConstraint object.

File: modules/contact/src/AugmentedLagrangianContactProblem.C
Comment (bwspenc) [reply modules/contact/src/AugmentedLagrangianContactProblem.C]: I think it would be better for these parameters to be defined for the individual contact constraints (penetration_tolerance, tangential_increment_stick_tolerance, tangential_friction_force_tolerance). I think I'd change their names too:
penetration_tolerance -> al_penetration_tolerance
tangential_increment_stick_tolerance -> al_incremental_slip_tolerance
tangential_friction_force_tolerance -> al_frictional_force_tolerance
They also need better descriptions.

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Comment by jiangwen84** [comment]
@bwspenc I made all the changes according to your comments. Once you think input syntax is good, @acasagran could start converting the remaining contact tests.

**Comment by jiangwen84** [comment]
I just pushed @acasagran 's commit of adding new AL tests.

File: modules/contact/src/AugmentedLagrangianContactProblem.C
Comment (bwspenc) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: remove commented out code

File: modules/contact/src/AugmentedLagrangianContactProblem.C
Comment (bwspenc) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: Get rid of the extra space after the newline.  Also, maybe get rid of the second newline.

File: modules/contact/src/AugmentedLagrangianContactProblem.C
Comment (bwspenc) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: No braces around one-liners

File: modules/contact/src/ContactAction.C
Comment (bwspenc) [inline modules/contact/src/ContactAction.C]: Maybe you need to get rid of the default here

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: delete blank line

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: delete blank line

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: This stuff won't go to the console if you hit it on a processor other than processor 0.

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: You could print the appropriate message based on the value of `converged` here, and then it would go to processor 0's output.

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: delete commented code

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: I'm not sure that making these fuzzy here is the right thing to do.  `capacity` is set to exactly zero if it should be zero.  Was this important for getting a model to work correctly?

File: modules/contact/src/MechanicalContactConstraint.C
Comment (bwspenc) [inline modules/contact/src/MechanicalContactConstraint.C]: delete

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/contact/src/MechanicalContactConstraint.C
Comment (jiangwen84) [reply modules/contact/src/MechanicalContactConstraint.C]: I forgot this one. I will fix it now.

**Comment by jiangwen84** [comment]
@bwspenc I fixed the penalty_slip issue.

**Review by bwspenc — APPROVED** [review]

**Comment by brianmoose** [comment]
Looks like `frictional_02_aug.i` has unused parameters and is causing heavy tests to fail: https://www.moosebuild.org/job/130035/

**Comment by jiangwen84** [comment]
@brianmoose Thanks. We will fix it (#10165).

**Comment by brianmoose** [comment]
It also looks like this causes a problem with PBS testing: https://www.moosebuild.org/job/130033/
The fix might be as simple as changing the output base for one of the tests but @milljm might have other ideas.

**Comment by jiangwen84** [comment]
The PBS one seems a tolerance issue. @acasagran might have better idea about this one.

**Comment by brianmoose** [comment]
The only reason I thought it might be an output issue is that it only fails in PBS and PBS does
things a little different than regular testing. Another test could overwrite the output file in between when the original test runs and when the csvdiff takes place.
But I will leave it up to you guys...

**Comment by brianmoose** [comment]
Is anybody looking into the PBS failure that this PR caused? https://www.moosebuild.org/job/131491/

**Comment by bwspenc** [comment]
Pull request #10214, which got merged yesterday, should have fixed that test.

**Comment by brianmoose** [comment]
Yes, everything looks good now, thanks!

**Pull request merged**
